File: LeetCode_POTD/June_2026/June_16.java
package Leetcode_POTD.June_2026;

import java.util.ArrayDeque;

/* 
Just did as mentioned: Use StringBUilder to prevent excessive copying of string everytime 
Better method: processStr1(Deque + reverse flag + StringBuilder)
*/
public class June_16 {
    public String processStr(String s) {
        StringBuilder res=new StringBuilder();
        for(int i=0;i<s.length();i++){
            if(s.charAt(i)=='*'){
                if(res.length()>0) res.deleteCharAt(res.length() - 1); 
            }
            else if(s.charAt(i)=='#'){
                res.append(res);
            }
            else if(s.charAt(i)=='%'){
                res.reverse();
            }
            else{
                res.append(s.charAt(i));
            }
        }
        return res.toString();
    }

    public String processStr1(String s) {
        ArrayDeque<Character> dq = new ArrayDeque<>();
        boolean rev = false;
        for (char ch : s.toCharArray()) {
            if (ch == '%') {
                rev = !rev;                          // O(1) — just flip flag
            } else if (ch == '#') {
                Character[] copy = dq.toArray(new Character[0]);
                for (char c : copy) dq.addLast(c);  // O(n) — unavoidable
            } else if (ch == '*') {
                if (!dq.isEmpty()) {
                    if (rev) dq.pollFirst();          // O(1)
                    else     dq.pollLast();           // O(1)
                }
            } else {
                if (rev) dq.addFirst(ch);            // O(1)
                else     dq.addLast(ch);             // O(1)
            }
        }
        // Final output: respect the rev flag
        StringBuilder sb = new StringBuilder();
        if (rev) {
            Iterator<Character> it = dq.descendingIterator();
            while (it.hasNext()) sb.append(it.next());
        } else {
            for (char c : dq) sb.append(c);
        }
        return sb.toString();
    }
}

File: LeetCode_POTD/June_2026/June_16.py
"""
Just did as mentioned: Using list as string are mutable(But on LC using string gave lesser runtime-I think may be because of small constraint(len(str)<=20))
Better method: processStr1(Deque + reverse flag + List)
"""

from collections import deque

class Solution:
    def processStr(self, s: str) -> str:
        res=[]
        for i in s:
            if i =='*':
                if res: res.pop()
            elif i=='#':
                res.extend(res)
            elif i=='%':
                res.reverse()
            else:
                res.append(i)
        return "".join(res)
    
    def processStr1(s: str) -> str:
        dq = deque()
        rev = False
        for ch in s:
            if ch == '%':
                rev = not rev                        # O(1) — just flip flag
            elif ch == '#':
                dq.extend(list(dq))                  # O(n) — unavoidable
            elif ch == '*':
                if dq:
                    if rev: dq.popleft()             # O(1)
                    else:   dq.pop()                 # O(1)
            else:  # lowercase letter
                if rev: dq.appendleft(ch)            # O(1)
                else:   dq.append(ch)                # O(1)
        # Final output: respect the rev flag
        return ''.join(reversed(dq)) if rev else ''.join(dq)